### ISSUE TITLE
Wrong Number of Labels for Train

### ISSUE THREAD

**Issue by loretoparisi** [facebookresearch/fastText#359]
In my dataset the number of labels is **229**, where labels start with the `__label__` prefix.
When I run training with the following parameters

```
  '-input',
  '/var/folders/_b/szqwdfn979n4fdg7f2j875_r0000gn/T/3f9453e5-8c55-41db-99a8-48e97cc9e164.csv',
  '-output',
  '/root/my_model',
  '-wordNgrams',
  '2',
  '-minCount',
  1,
  '-minCountLabel',
  1,
  '-minn',
  3,
  '-maxn',
  6,
  '-t',
  0.0001,
  '-bucket',
  10000000,
  '-dim',
  '10',
  '-lr',
  '0.1',
  '-ws',
  15,
  '-loss',
  'ns',
  '-lrUpdateRate',
  100,
  '-epoch',
  '20',
  '-thread',
  4,
  '-verbose',
  4 ]
```

I get

```
Read 38M words
Number of words:  165190
Number of labels: 217
```

so some labels are missing. I'm sure of the labels count, since I get the same result using like:

```bash
#!/bin/bash
COL=$1
FILE=$2
SEP=$3
awk -v col="$COL" -F $"$SEP" '   {c[$col]++}
                 END{
                     p=1; for (i in c) { printf("%s %s'$SEP'%s\n",p,i,c[i]);p++ }
                 }' $FILE
```

that outputs like

```
...
1 __label__classicalorchestral	1
2 __label__latinlatinjazz	3
3 __label__rbsoulfunk	36
...
227 __label__countryhonkytonk	17
228 __label__blueschicagoblues	7
229 __label__popoldies	15
```

so the count is correct: 229 different values in the first column.

**Comment by EdouardGrave** [comment]
Hi @loretoparisi,

Thank you for reporting this.

How many labels appear only once in your dataset? If using the Python bindings, could you use the `get_labels` function to find which labels are missing in the fastText model?

Thank you for your help!
Edouard

**Comment by loretoparisi** [comment]
Hey @EdouardGrave thanks for the suggestions. First of all I have adjusted a bit my counter script 'cause I was missing a separator:

```bash
COL=$1
FILE=$2
SEP=$3
awk -v col="$COL" -F $"$SEP" '   {c[$col]++}
                 END{
                     p=1; for (i in c) { printf("%s'$SEP'%s'$SEP'%s\n",p,i,c[i]);p++ }
                 }' $FILE
```
then I did 

```
$ ./count_classes.sh 1 /root/my_dataset.csv "\t" | awk '$3==1 { printf "%s\t%s\n",$2,$3 }'
__label__classicalorchestral	1
__label__countryoutlawcountry	1
__label__worldsoca	1
__label__popmalaysianpop	1
__label__jazzcrossoverjazz	1
__label__indianindianfolk	1
__label__worldindonesianreligious	1
__label__worldfarsi	1
__label__latinraices	1
__label__worldtraditionalceltic	1
__label__newageenvironmental	1
__label__holidaychristmasrock	1
__label__poppinoypop	1
__label__arabic	1
__label__indianindianclassical	1
__label__holidaychristmasclassic	1
__label__worldflamenco	1
__label__worlddangdut	1
__label__classicalopera	1
__label__popcpop	1
```

and in Python:

```python
>>> f = load_model('/root/my_model.bin')
>>> labels = f.get_labels()
>>> len(labels)
217
```

so we have 229-20

```
$ ./count_classes.sh 1 /root/my_dataset.csv "\t" | awk '$3==1 { printf "%s\t%s\n",$2,$3 }' | wc -l
      20
```

**Comment by loretoparisi** [comment]
**[UPDATE]**
So I came up with this:

```
./count_classes.sh 1 /root/my_dataset.csv "\t" | awk '{ printf "%s,%s\n",$2,$3 }' > /root/labels.csv
```

and then in python if I'm not doing anything wrong:

```python
>>> import fastText
>>> from fastText import load_model
>>> f = load_model('/root/my_model.bin')
>>> labels = f.get_labels()
>>> import csv
>>> with open('/root/labels.csv') as f:
>>>...d = dict(filter(None, csv.reader(f)))
>>> l=[k for k,v in d.items()]
>>> len(labels)
217
>>> len(l)
229
>>> missing=list(set(l) - set(labels))
>>> len(missing)
13
>>> missing
['__label__classicalopera', '__label__worldflamenco', '__label__popmalaysianpop', '__label__classicalorchestral', '__label__popcpop', '__label__worldjapanesepop', '__label__popcantopophkpop', '__label__latinraices', '__label__indianindianfolk', '__label__countrycountrygospel', '__label__worlddangdut', '__label__worldindonesianreligious', '__label__rockarenarock']

**Comment by EdouardGrave** [comment]
Hi @loretoparisi,

it seems that only low frequency labels are missing. This is probably happening because the vocabulary is thresholded during building. One thing that you can try is to "copy" each label multiple time (three times should work here), so that they do not get filtered out during dictionary building. This will not change the learning process. If this solve the problem, we will try to fix the issue in the dictionary building.

Thank you for your help!

**Comment by loretoparisi** [comment]
Hey @EdouardGrave It seems that you are right. This is what happens:

```
Read 76M words{ W: 240267, L: 229 }
Read 76M words
Number of words:  240267
Number of labels: 229
```

and

```
>>> import fastText
>>> from fastText import load_model
>>> f = load_model('/var/folders/_b/szqwdfn979n4fdg7f2j875_r0000gn/T/genre_model.bin')
>>> labels = f.get_labels()
>>> len(labels)
229
>>> import csv
>>> with open('/root/labels.csv') as f:
...     d = dict(filter(None, csv.reader(f)))
...     l=[k for k,v in d.items()]
... 
>>> len(l)
229
>>> missing=list(set(l) - set(labels))
>>> len(missing)
0
```

after I had augmented the data adding at least 3 occurrences of the `missing` labels.

**Comment by cpuhrsch** [comment]
Hello @loretoparisi,

Thank you for your post and for being an active member of our community. I'm closing this issue now, but please feel free to reopen it at any time if you do not consider it resolved.

Thanks,
Christian

**Comment by loretoparisi** [comment]
@cpuhrsch you are  welcome! Did you have the opportunity to check the vocabulary threshold settings?

**Comment by cpuhrsch** [comment]
Hello @loretoparisi,

I'll note it down as a new task :)

Thanks,
Christian
